examples/OAuthConsumer/Program.cs

```diff
@@ -0,0 +1,192 @@
+            }
+
+
+            using (var consumer = new ConsumerBuilder<string, string>(consumerConfig)
+                                    .SetOAuthBearerTokenRefreshHandler(OauthCallback).Build())
+            {
+                Console.WriteLine("\n-----------------------------------------------------------------------");
+                Console.WriteLine($"Consumer {consumer.Name} consuming from topic {topicName}.");
+                Console.WriteLine("-----------------------------------------------------------------------");
+                Console.WriteLine("Ctrl-C to quit.\n");
+
+                consumer.Subscribe(topicName);
+                CancellationTokenSource cts = new CancellationTokenSource();
+                Console.CancelKeyPress += (_, e) =>
+                {
+                    e.Cancel = true; // prevent the process from terminating.
+                    cts.Cancel();
+                };
+
+                try
+                {
+                    while (true)
+                    {
+                        try
+                        {
+                            var consumeResult = consumer.Consume(cts.Token);
+
+                            Console.WriteLine($"Received message at {consumeResult.TopicPartitionOffset}: {consumeResult.Message.Value}");
+                            try
+                            {
+                                consumer.StoreOffset(consumeResult);
+                            }
+                            catch (KafkaException e)
+                            {
+                                Console.WriteLine($"Store Offset error: {e.Error.Reason}");
+                            }
+                        }
+                        catch (ConsumeException e)
+                        {
+                            Console.WriteLine($"Consume error: {e.Error.Reason}");
+                        }
+                    }
+                }
+                catch (OperationCanceledException)
+                {
+                    Console.WriteLine("Closing consumer.");
+                    consumer.Close();
+                }
+            }
+        }
+
+        private static string ToUnpaddedBase64(string s)
+            => Convert.ToBase64String(Encoding.UTF8.GetBytes(s)).TrimEnd('=');
+
+        private static OAuthBearerToken retrieveUnsecuredToken(String oauthConfig)
+        {
+            Console.WriteLine("Refreshing the token");
+
+            var parsedConfig = new Dictionary<String, String>();
+            foreach (Match match in Regex.Matches(oauthConfig, OauthconfigKeyValueRegexPattern))
+            {
+                parsedConfig[match.Groups[1].ToString()] = match.Groups[2].ToString();
+            }
+
+            if (!parsedConfig.ContainsKey(PrincipalKey) || !parsedConfig.ContainsKey(ScopeKey) || parsedConfig.Count > 2)
+            {
+                Console.WriteLine($"Invalid OAuth config {oauthConfig} passed. Ignoring the token refresh");

```

Comment:
The proper way to handle this is to raise an error which is then passed to SetTokenFailure() in the refresh delegate.

Reply:
Makes sense, throwing an exception now and setting `OAuthBearerSetTokenFailure` on catching the exception.